docs/deprecation-process.md

```diff
@@ -0,0 +1,309 @@
+
+In any case, never implement the deprecation and the removal in the same
+release: wait.
+
+### Prepare removal
+
+Implement the removal issue with a PR.
+
+Remove all the deprecated code.
+
+Remove all the tooling (compiling options) related to the deprecated code,
+if any.
+
+Remove all the relevant entries in `DEPRECATED`.
+
+Add a `CHANGELOG` entry for the removal.
+
+Get the removal PR reviewed and merged to the main branch.
+
+### Release removal
+
+When making a new opentelemetry-cpp release,
+in addition to the regular release tasks:
+
+* In the `CHANGELOG` and the release notes, add references
+  to the new removed items for this release.
+
+## Technical guidelines
+
+### CMake options
+
+Assume the option `WITH_FOO` needs to be deprecated.
+
+Code using `WITH_FOO=OFF` or `WITH_FOO=ON` must build as before,
+yet users should be notified if they use `WITH_FOO` in their build.
+
+Assume a build option `WITH_NO_DEPRECATION` already exists.
+
+In a normal build, used in production, code is compiled with
+`WITH_NO_DEPRECATION=OFF`.
+
+In a verification build, code is compiled with `WITH_NO_DEPRECATION=ON`.
+
+Implement the following logic in CMake:
+
+```cmake
+  option(WITH_FOO "DEPRECATED - With the foo feature" OFF)
+
+  if(WITH_FOO)
+    if(WITH_NO_DEPRECATION)
+      message(FATAL_ERROR "WITH_FOO is deprecated")
+    else()
+      message(WARNING "WITH_FOO is deprecated")
+    endif()
+  endif
+```
+
+The verification build is not meant to be used in production.
+It is meant to find all references of deprecated items.
+
+If the verification build succeeds, the user code is guaranteed to be clean,
+and will not be impacted by the removal to come.
+
+When implementing such logic, document it in the mitigation section,
+in file `DEPRECATED`.
+
+### C++ deprecation
+
+Assume a C++ item needs to be deprecated.
+
+For example:
+
+```cpp
+  struct some_options
+  {
+    int x;
+    int y; // TODO: deprecate me
+    int z;
+  };
+```
+
+Code using y must build as before, and yet users should be notified if still
+using y.
+
+Implement the following change:
+
+```cpp
+  struct some_options
+  {
+    int x;
+#ifdef HAVE_DEPRECATED_CODE
+    int y; // deprecated, to be removed
+#endif /* HAVE_DEPRECATED_CODE */
+    int z;
+  };
+```
+
+A regular build, to use in production, will build with
+`HAVE_DEPRECATED_CODE` defined.
+
+A verification build, to _not_ use in production,
+will build with `HAVE_DEPRECATED_CODE` undefined.
+
+Even if the verification build succeeds, it really should not be used in
+production, because the memory layout just changed, breaking the ABI.
+
+This verification build is used to enforce:
+
+* the API (Application _Programming_ Interface), is unchanged,
+  aka the source code will build before and after the removal.
+
+By the time the removal is implemented:
+
+```cpp
+  struct some_options
+  {
+    int x;
+    int z;
+  };
+```
+
+the new release will have:
+
+* an API still compatible
```

Comment:
Please correct me, but it won't be API compatible if the public methods / global functions are going to be removed ?

Reply:
Will clarify and reword for clarity.